### ISSUE TITLE
Provisioning does not take into account spot availability in AZ

### ISSUE THREAD

**Issue by mKeRix** [aws/karpenter#2243]
## Version

**Karpenter**: v0.13.2

**Kubernetes**: v1.22.9

## Expected Behavior

Karpenter does not include instance types in CreateFleet calls if they are not supported as spot in the availability zone it chooses.

## Actual Behavior

Karpenter sends off a CreateFleet request including the unavailable instance type, which results in an `InvalidFleetConfiguration` error. It will retry the provisioning in a loop, however this will always fail as it keeps using the same instance types in the call.

I've checked the region using the EC2 describe APIs and can confirm that the instance type is offered in all availability zones (describe-instance-type-offerings), and as spot (describe-instance-types). Some further searching led me to the spot pricing history, where I can see that the instance type only shows prices for two of the three zones. In our case, eu-central-1a (euc1-az2) does not have a price history, which leads me to believe that this instance type is not offered as spot in that AZ.  

## Steps to Reproduce the Problem

1. Deploy a cluster with Karpenter in eu-central-1.
2. Create a provisioner with spot instances enabled.
3. Provision a workload by limiting its nodeSelector to only run in whichever availability zone name maps to the zone id `euc1-az2` in your account, as well as the instance type to r6i.32xlarge.
4. Observe Karpenter logs and pod not being able to come up.

Note that this issue can happen without any of the constraints depending on all the factors that play into scheduling, but this test setup should be consistently reproducable at least.

## Resource Specs and Logs

Karpenter logs:

```
ERROR controller.provisioning Provisioning failed, launching node, creating cloud provider instance, with fleet error(s), InvalidFleetConfiguration: Your requested instance type (r6i.32xlarge) is not supported in your requested Availability Zone (eu-central-1a). {"commit": "c2372b2"}
```

CloudTrail (shortened to remove private information):

key | value
-- | --
responseElements.CreateFleetResponse.errorSet.item.errorCode | InvalidFleetConfiguration
responseElements.CreateFleetResponse.errorSet.item.errorMessage | Your requested instance type (r6i.32xlarge) is not supported in your requested Availability Zone (eu-central-1a).
responseElements.CreateFleetResponse.errorSet.item.launchTemplateAndOverrides.launchTemplateSpecification.launchTemplateId | lt-xyz
responseElements.CreateFleetResponse.errorSet.item.launchTemplateAndOverrides.launchTemplateSpecification.version | 1
responseElements.CreateFleetResponse.errorSet.item.launchTemplateAndOverrides.overrides.availabilityZone | eu-central-1a
responseElements.CreateFleetResponse.errorSet.item.launchTemplateAndOverrides.overrides.instanceType | r6i.32xlarge
responseElements.CreateFleetResponse.errorSet.item.launchTemplateAndOverrides.overrides.priority | 1
responseElements.CreateFleetResponse.errorSet.item.launchTemplateAndOverrides.overrides.subnetId | subnet-xyz
responseElements.CreateFleetResponse.errorSet.item.lifecycle | spot
responseElements.CreateFleetResponse.fleetId | fleet-xyz
responseElements.CreateFleetResponse.requestId | xyz
responseElements.CreateFleetResponse.xmlns | http://ec2.amazonaws.com/doc/2016-11-15/

Provisioner:

<details>

```yaml
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  labels:
    idealo.de/subnet-type: private
  limits:
    resources:
      cpu: 1k
  providerRef:
    name: private-node
  requirements:
    - key: karpenter.sh/capacity-type
      operator: In
      values:
        - spot
        - on-demand
    - key: kubernetes.io/arch
      operator: In
      values:
        - amd64
        - arm64
    - key: node.kubernetes.io/instance-type
      operator: In
      values:
        - a1.2xlarge
        - a1.4xlarge
        - a1.xlarge
        - c5.12xlarge
        - c5.18xlarge
        - c5.24xlarge
        - c5.2xlarge
        - c5.4xlarge
        - c5.9xlarge
        - c5.xlarge
        - c5a.12xlarge
        - c5a.16xlarge
        - c5a.24xlarge
        - c5a.2xlarge
        - c5a.4xlarge
        - c5a.8xlarge
        - c5a.xlarge
        - c5ad.12xlarge
        - c5ad.16xlarge
        - c5ad.24xlarge
        - c5ad.2xlarge
        - c5ad.4xlarge
        - c5ad.8xlarge
        - c5ad.xlarge
        - c5d.12xlarge
        - c5d.18xlarge
        - c5d.24xlarge
        - c5d.2xlarge
        - c5d.4xlarge
        - c5d.9xlarge
        - c5d.xlarge
        - c5n.18xlarge
        - c5n.2xlarge
        - c5n.4xlarge
        - c5n.9xlarge
        - c5n.xlarge
        - c6a.12xlarge
        - c6a.16xlarge
        - c6a.24xlarge
        - c6a.2xlarge
        - c6a.32xlarge
        - c6a.4xlarge
        - c6a.8xlarge
        - c6a.xlarge
        - c6g.12xlarge
        - c6g.16xlarge
        - c6g.2xlarge
        - c6g.4xlarge
        - c6g.8xlarge
        - c6g.xlarge
        - c6gd.12xlarge
        - c6gd.16xlarge
        - c6gd.2xlarge
        - c6gd.4xlarge
        - c6gd.8xlarge
        - c6gd.xlarge
        - c6gn.12xlarge
        - c6gn.16xlarge
        - c6gn.2xlarge
        - c6gn.4xlarge
        - c6gn.8xlarge
        - c6gn.xlarge
        - c6i.12xlarge
        - c6i.16xlarge
        - c6i.24xlarge
        - c6i.2xlarge
        - c6i.32xlarge
        - c6i.4xlarge
        - c6i.8xlarge
        - c6i.xlarge
        - d3.2xlarge
        - d3.4xlarge
        - d3.8xlarge
        - d3.xlarge
        - g4ad.16xlarge
        - g4ad.2xlarge
        - g4ad.4xlarge
        - g4ad.8xlarge
        - g4ad.xlarge
        - g4dn.12xlarge
        - g4dn.16xlarge
        - g4dn.2xlarge
        - g4dn.4xlarge
        - g4dn.8xlarge
        - g4dn.xlarge
        - g5.12xlarge
        - g5.16xlarge
        - g5.24xlarge
        - g5.2xlarge
        - g5.4xlarge
        - g5.8xlarge
        - g5.xlarge
        - i3en.12xlarge
        - i3en.24xlarge
        - i3en.2xlarge
        - i3en.3xlarge
        - i3en.6xlarge
        - i3en.xlarge
        - i4i.16xlarge
        - i4i.2xlarge
        - i4i.32xlarge
        - i4i.4xlarge
        - i4i.8xlarge
        - i4i.xlarge
        - inf1.24xlarge
        - inf1.2xlarge
        - inf1.6xlarge
        - inf1.xlarge
        - m5.12xlarge
        - m5.16xlarge
        - m5.24xlarge
        - m5.2xlarge
        - m5.4xlarge
        - m5.8xlarge
        - m5.xlarge
        - m5a.12xlarge
        - m5a.16xlarge
        - m5a.24xlarge
        - m5a.2xlarge
        - m5a.4xlarge
        - m5a.8xlarge
        - m5a.xlarge
        - m5ad.12xlarge
        - m5ad.16xlarge
        - m5ad.24xlarge
        - m5ad.2xlarge
        - m5ad.4xlarge
        - m5ad.8xlarge
        - m5ad.xlarge
        - m5d.12xlarge
        - m5d.16xlarge
        - m5d.24xlarge
        - m5d.2xlarge
        - m5d.4xlarge
        - m5d.8xlarge
        - m5d.xlarge
        - m5dn.12xlarge
        - m5dn.16xlarge
        - m5dn.24xlarge
        - m5dn.2xlarge
        - m5dn.4xlarge
        - m5dn.8xlarge
        - m5dn.xlarge
        - m5n.12xlarge
        - m5n.16xlarge
        - m5n.24xlarge
        - m5n.2xlarge
        - m5n.4xlarge
        - m5n.8xlarge
        - m5n.xlarge
        - m5zn.12xlarge
        - m5zn.2xlarge
        - m5zn.3xlarge
        - m5zn.6xlarge
        - m5zn.xlarge
        - m6a.12xlarge
        - m6a.16xlarge
        - m6a.24xlarge
        - m6a.2xlarge
        - m6a.32xlarge
        - m6a.4xlarge
        - m6a.8xlarge
        - m6a.xlarge
        - m6g.12xlarge
        - m6g.16xlarge
        - m6g.2xlarge
        - m6g.4xlarge
        - m6g.8xlarge
        - m6g.xlarge
        - m6gd.12xlarge
        - m6gd.16xlarge
        - m6gd.2xlarge
        - m6gd.4xlarge
        - m6gd.8xlarge
        - m6gd.xlarge
        - m6i.12xlarge
        - m6i.16xlarge
        - m6i.24xlarge
        - m6i.2xlarge
        - m6i.32xlarge
        - m6i.4xlarge
        - m6i.8xlarge
        - m6i.xlarge
        - p4d.24xlarge
        - r5.12xlarge
        - r5.16xlarge
        - r5.24xlarge
        - r5.2xlarge
        - r5.4xlarge
        - r5.8xlarge
        - r5.xlarge
        - r5a.12xlarge
        - r5a.16xlarge
        - r5a.24xlarge
        - r5a.2xlarge
        - r5a.4xlarge
        - r5a.8xlarge
        - r5a.xlarge
        - r5ad.12xlarge
        - r5ad.16xlarge
        - r5ad.24xlarge
        - r5ad.2xlarge
        - r5ad.4xlarge
        - r5ad.8xlarge
        - r5ad.xlarge
        - r5b.12xlarge
        - r5b.16xlarge
        - r5b.24xlarge
        - r5b.2xlarge
        - r5b.4xlarge
        - r5b.8xlarge
        - r5b.xlarge
        - r5d.12xlarge
        - r5d.16xlarge
        - r5d.24xlarge
        - r5d.2xlarge
        - r5d.4xlarge
        - r5d.8xlarge
        - r5d.xlarge
        - r5dn.12xlarge
        - r5dn.16xlarge
        - r5dn.24xlarge
        - r5dn.2xlarge
        - r5dn.4xlarge
        - r5dn.8xlarge
        - r5dn.xlarge
        - r5n.12xlarge
        - r5n.16xlarge
        - r5n.24xlarge
        - r5n.2xlarge
        - r5n.4xlarge
        - r5n.8xlarge
        - r5n.xlarge
        - r6a.12xlarge
        - r6a.16xlarge
        - r6a.24xlarge
        - r6a.2xlarge
        - r6a.32xlarge
        - r6a.4xlarge
        - r6a.8xlarge
        - r6a.xlarge
        - r6g.12xlarge
        - r6g.16xlarge
        - r6g.2xlarge
        - r6g.4xlarge
        - r6g.8xlarge
        - r6g.xlarge
        - r6gd.12xlarge
        - r6gd.16xlarge
        - r6gd.2xlarge
        - r6gd.4xlarge
        - r6gd.8xlarge
        - r6gd.xlarge
        - r6i.12xlarge
        - r6i.16xlarge
        - r6i.24xlarge
        - r6i.2xlarge
        - r6i.32xlarge
        - r6i.4xlarge
        - r6i.8xlarge
        - r6i.xlarge
        - t3.2xlarge
        - t3.xlarge
        - t3a.2xlarge
        - t3a.xlarge
        - t4g.2xlarge
        - t4g.xlarge
        - x2idn.16xlarge
        - x2idn.24xlarge
        - x2idn.32xlarge
        - x2iedn.16xlarge
        - x2iedn.24xlarge
        - x2iedn.2xlarge
        - x2iedn.32xlarge
        - x2iedn.4xlarge
        - x2iedn.8xlarge
        - x2iedn.xlarge
        - z1d.12xlarge
        - z1d.2xlarge
        - z1d.3xlarge
        - z1d.6xlarge
        - z1d.xlarge
  startupTaints:
    - effect: NoExecute
      key: node.cilium.io/agent-not-ready
      value: 'true'
  ttlSecondsAfterEmpty: 60
  ttlSecondsUntilExpired: 604800
```

</details>

**Comment by dewjam** [comment]
Hey there @mKeRix ,
Thanks for reporting this issue.  I'll be digging into this shortly.

**Comment by dewjam** [comment]
Hey @mKeRix .  The r6i.32xlarge instance type appears to be available as spot in eu-central-1a, but that specific instance type is capacity constrained.  So that might explain why it's listed as an offering, but still not able to be launched in eu-central-1a.

The one part I'm still confused by is you are seeing Karpenter continue to retry the failed instance type.  My understanding is Karpenter should mark the instance type as unavailable in that AZ for a period of time as to prevent continuous retry.  

https://github.com/aws/karpenter/blob/24694bc6a3862c00fbcdbebf3ebe19fe4d5daae5/pkg/cloudprovider/aws/instancetypes.go#L103

**Comment by mKeRix** [comment]
This is the result we get for eu-central-1a:

```shell
$ aws ec2 describe-spot-price-history --availability-zone eu-central-1a --filters Name=instance-type,Values=r6i.32xlarge Name=product-description,Values=Linux/UNIX
{
    "SpotPriceHistory": []
}
```

This is how the results look for us in eu-central-1b and eu-central-1c (shortened to only the first item in the array):

```shell
$ aws ec2 describe-spot-price-history --availability-zone eu-central-1b --filters Name=instance-type,Values=r6i.32xlarge Name=product-description,Values=Linux/UNIX
{
    "SpotPriceHistory": [
        {
            "AvailabilityZone": "eu-central-1b",
            "InstanceType": "r6i.32xlarge",
            "ProductDescription": "Linux/UNIX",
            "SpotPrice": "2.450800",
            "Timestamp": "2022-08-02T23:30:33+00:00"
        },
}
```

The following two commands report it as available as spot and in all AZs however:

```
$ aws ec2 describe-instance-types --filters Name=instance-type,Values=r6i.32xlarge Name=supported-usage-class,Values=spot
$ aws ec2 describe-instance-type-offerings --location-type availability-zone --filters Name=location,Values=eu-central-1a Name=instance-type,Values=r6i.32xlarge
```

We also saw that this pattern seems to apply to all instance types in the r6i family (e.g. r6i.4xlarge). 

Searching through the code for this specific cache I believe that the `InvalidFleetConfiguration` error we saw would not lead to it being put into that cache, based on the error codes that it checks: https://github.com/aws/karpenter/blob/37933eab2d8042fcb23f22802faf1ec3766985b9/pkg/cloudprovider/aws/errors.go#L35-L41

If it's of any help, I'm happy to submit an official ticket to the AWS support to get the exact reasoning for this error as well.

**Comment by dewjam** [comment]
ah, yes, you're right about the error @mKeRix .  I've reached out to some resources internally, but you're certainly welcome to open an official ticket as well.

**Comment by dewjam** [comment]
A quick update @mKeRix .  We are working with some teams internally to ensure the data returned from the `describe-instance-type-offerings` call is up to date.  This should solve the immediate problem in the eu-central-1 zone you mentioned.  

Also, as @bwagner5 mentioned, we are looking into using spot price history when determining which instance types are available in availability zones (#2256).  This will hopefully allow Karpenter to work around this issue in the future (and give Karpenter access to more accurate spot pricing info).

**Comment by jonathan-innis** [comment]
Should be resolved with #2283 as we now rely on the spot pricing info to give us the relevant AZs where a capacity type is available
